libr/bin/dwarf.c

```diff
@@ -1319,321 +1519,523 @@ static void r_bin_dwarf_dump_debug_info(FILE *f, const RBinDwarfDebugInfo *inf)
 
-	if (cu->hdr.length > debug_str_len) {
-		//avoid oob read
-		return NULL;
+		buf = parse_attr_value (buf, buf_end - buf, &abbrev->defs[i],
+			&die->attr_values[i], hdr, debug_str, debug_str_len);
+
+		RBinDwarfAttrValue *attribute = &die->attr_values[i];
+
+		bool is_valid_string_form = (attribute->attr_form == DW_FORM_strp ||
+						    attribute->attr_form == DW_FORM_string) &&
+			attribute->string.content;
+		// TODO  does this have a purpose anymore?
+		// Or atleast it needs to rework becase there will be
+		// more comp units -> more comp dirs and only the last one will be kept
+		if (attribute->attr_name == DW_AT_comp_dir && is_valid_string_form) {
+			const char *name = attribute->string.content;
+			sdb_set (sdb, "DW_AT_comp_dir", name, 0);
+		}
+		die->length++;
 	}
-	while (buf && buf < buf_end && buf >= obuf) {
-		if (cu->length && cu->capacity == cu->length) {
-			r_bin_dwarf_expand_cu (cu);
+
+	return buf;
+}
+
+/**
+ * @brief Reads throught comp_unit buffer and parses all its DIEntries
+ * 
+ * @param sdb
+ * @param buf_start Start of the compilation unit data
+ * @param unit Unit to store the newly parsed information
+ * @param abbrevs Parsed abbrev section info of *all* abbreviations
+ * @param abbr_offset index offset into Abbrev array to current comp unit abbrevs
+ * @param debug_str Ptr to string section start
+ * @param debug_str_len Length of the string section
+ * 
+ * @return const ut8* Update buffer
+ */
+static const ut8 *parse_comp_unit(Sdb *sdb, const ut8 *buf_start,
+		RBinDwarfCompUnit *unit, const RBinDwarfDebugAbbrev *abbrevs,
+		size_t abbr_offset, const ut8 *debug_str, size_t debug_str_len) {
+
+	const ut8 *buf = buf_start;
+	const ut8 *buf_end = buf_start + unit->hdr.length - unit->hdr.header_size;
+	size_t i;
+
+	while (buf && buf < buf_end && buf >= buf_start) {
+		if (unit->length && unit->capacity == unit->length) {
+			expand_cu (unit);
 		}
+		RBinDwarfDie *die = &unit->dies[unit->length];
+		// DIE starts with ULEB128 with the abbreviation code
+		ut64 abbr_code;
 		buf = r_uleb128 (buf, buf_end - buf, &abbr_code);
-		if (abbr_code > da->length || !buf || buf >= buf_end) {
-			return NULL;
-		}
 
-		r_bin_dwarf_init_die (&cu->dies[cu->length]);
+		init_die (die, abbr_code);
+
+		if (abbr_code > abbrevs->length || !buf || buf >= buf_end) {
+			unit->length++; // we wanna store this entry too, usually the last one is null_entry
+			return buf; // return the buffer to parse next compilation units
+		}
+		// there can be "null" entries that have abbr_code == 0
 		if (!abbr_code) {
-			cu->dies[cu->length].abbrev_code = 0;
-			cu->length++;
-			buf++;
+			unit->length++;
 			continue;
 		}
+		ut64 abbr_idx = abbr_code + abbr_offset;
+
+		if (abbrevs->length < abbr_idx) {
+			return NULL;
+		}
 
-		cu->dies[cu->length].abbrev_code = abbr_code;
-		cu->dies[cu->length].tag = da->decls[abbr_code - 1].tag;
-		abbr_code += offset;
+		RBinDwarfAbbrevDecl *abbrev = &abbrevs->decls[abbr_idx - 1];
```

Comment:
Why are you using `abbr_idx` to access the `abbrevs->decls` array? Shouldn't it be `abbr_code`? Also, `abbr_idx` is not checked and it could go out of bounds.

Reply:
It is checked right below it

```c
		ut64 abbr_idx = abbr_code + abbr_offset;

		if (abbrevs->length < abbr_idx) {
			return NULL;
		}
```

abbr_code is code of the abbreviation for the unit

> ut64 abbr_idx = abbr_code + abbr_offset;

abbr_idx is code + offset in the abbreviations array to the **current** compilation unit. Because all abbreviations from all compilation units are stored in the array and abbr_code is just for the current one.

Reply:
Ok, the entire comment above please :D

Reply:
Yeah sorry I realized it was ok after readng better the spec and looking at some dwarf dump output